packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/services/ComposedObjectService.java

```diff
@@ -0,0 +1,123 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Obeo.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.sirius.components.core.services;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.eclipse.sirius.components.core.api.IDefaultObjectService;
+import org.eclipse.sirius.components.core.api.IEditingContext;
+import org.eclipse.sirius.components.core.api.IObjectService;
+import org.eclipse.sirius.components.core.api.IObjectServiceDelegate;
+import org.springframework.stereotype.Service;
+
+/**
+ * Implementation of {@link IObjectService} which delegates to {@link IObjectServiceDelegate} or fallback to
+ * {@link IDefaultObjectService}.
+ *
+ * @author arichard
+ */
+@Service
+public class ComposedObjectService implements IObjectService {
+
+    private final List<IObjectServiceDelegate> objectServiceDelegates;
+
+    private final IDefaultObjectService defaultObjectService;
+
+    public ComposedObjectService(List<IObjectServiceDelegate> objectServiceDelegates, IDefaultObjectService defaultObjectService) {
+        this.objectServiceDelegates = Objects.requireNonNull(objectServiceDelegates);
+        this.defaultObjectService = Objects.requireNonNull(defaultObjectService);
+    }
+
+    @Override
+    public String getId(Object object) {
+        return this.objectServiceDelegates.stream()
+                .filter(delegate -> delegate.canHandle(null, object))
+                .findFirst()
+                .map(delegate -> delegate.getId(object))
+                .orElse(this.defaultObjectService.getId(object));
```

Comment:
This is not what we agreed on. In this case, if someone creates a bad delegate returning `null` then the default value would be returned instead and that's exactly what I don't want. If a delegates indicates that it can handle something, it has to handle it properly, no magic fallback. So the code should look like this:


```
var optionalDelegate = this.objectServiceDelegates.stream()
    .filter(delegate -> delegate.canHandle(object))
    .findFirst();
if (optionalDelegate.isPresent()) {
  return optionalDelegate.get().getId(object);
)
return this.defaultObjectService.getId(object);
```

When we will split the `IObjectService` interface into multiple interfaces like `ILabelService` etc, we could simplify the code a bit but it would keep the same philosophy. I fully expect the code to crash and burn as fast as possible if someone tells us that they can handle such an important responsibility and fail to fulfill it.

Reply:
Of course, this comment is valid for all the other parts of this code

Reply:
Done